source/common/network/io_uring_socket_handle_impl.cc

```diff
@@ -0,0 +1,415 @@
+#include "source/common/network/io_uring_socket_handle_impl.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+#include "envoy/event/dispatcher.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/common/assert.h"
+#include "source/common/common/utility.h"
+#include "source/common/io/io_uring_worker_impl.h"
+#include "source/common/network/address_impl.h"
+#include "source/common/network/io_socket_error_impl.h"
+#include "source/common/network/io_socket_handle_impl.h"
+#include "source/common/network/socket_interface_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+IoUringSocketHandleImpl::IoUringSocketHandleImpl(Io::IoUringWorkerFactory& io_uring_worker_factory,
+                                                 os_fd_t fd, bool socket_v6only,
+                                                 absl::optional<int> domain, bool is_server_socket)
+    : IoSocketHandleBaseImpl(fd, socket_v6only, domain),
+      io_uring_worker_factory_(io_uring_worker_factory),
+      io_uring_socket_type_(is_server_socket ? IoUringSocketType::Server
+                                             : IoUringSocketType::Unknown) {
+  ENVOY_LOG(trace, "construct io uring socket handle, fd = {}, type = {}", fd_,
+            ioUringSocketTypeStr());
+}
+
+IoUringSocketHandleImpl::~IoUringSocketHandleImpl() {
+  ENVOY_LOG(trace, "~IoUringSocketHandleImpl, type = {}", ioUringSocketTypeStr());
+  if (SOCKET_VALID(fd_)) {
+    // If the socket is owned by the main thread like a listener, it may outlive the IoUringWorker.
+    // We have to ensure that the current thread has been registered and the io_uring in the thread
+    // is still available.
+    // TODO(zhxie): for current usage of server socket and client socket, the check may be
+    // redundant.
+    if (io_uring_socket_type_ != IoUringSocketType::Unknown &&
+        io_uring_socket_type_ != IoUringSocketType::Accept &&
+        io_uring_worker_factory_.currentThreadRegistered() && io_uring_socket_.has_value()) {
+      if (io_uring_socket_->getStatus() != Io::IoUringSocketStatus::Closed) {
+        io_uring_socket_.ref().close(false);
+      }
+    } else {
+      // The TLS slot has been shut down by this moment with io_uring wiped out, thus better use the
```

Comment:
nit: Is it "better" to use close, or is it required?

Reply:
Required. rm `better`.